### PR TITLE
VideoCommon: add support for custom assets and a custom asset loader

### PR DESCRIPTION
**Pull request by iwubcode**
This splits out the assets + asset loader from #11681 to hopefully make it easier to review.

### CODE REVIEW

File: Source/Core/VideoCommon/GraphicsModSystem/Runtime/CustomAssetLoader.cpp [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+{
+void CustomAssetLoader::Init()
+{
+  m_file_monitor_thread_shutdown.Clear();
+
+  const size_t sys_mem = Common::MemPhysical();
+  const size_t recommended_min_mem = 2 * size_t(1024 * 1024 * 1024);
+  // keep 2GB memory for system stability if system RAM is 4GB+ - use half of memory in other cases
+  m_max_memory_available =
+      (sys_mem / 2 < recommended_min_mem) ? (sys_mem / 2) : (sys_mem - recommended_min_mem);
```
Comment (AdmiralCurtiss) [inline Source/Core/VideoCommon/GraphicsModSystem/Runtime/CustomAssetLoader.cpp:21]: This is copied from the existing custom texture logic, right? I assume the two will never be active at the same time?

**Comment by AdmiralCurtiss** [comment]
I think you need to add a few comments here (in the code, please) to explain how you intend these classes to be used. I get the idea that you want to create the object at a defined place and then later asynchronously add data to it, but the details are not entirely clear from just these classes. In particular, the logic in the CustomAssetLoader is kinda hard to follow.

A few more notes:
- What exactly is the `asset_id`, and why is it a `std::filesystem::path`? Does that make sense to be a `path` when we consider the possibility of synthesizing assets on the fly, or reading them from an archive without extracting/decompressing to disk?
- I also feel like having a mutex per asset is probably not a great idea. Though I suspect refactoring that out is probably somewhat difficult and definitely error-prone, so it's probably fine for now -- something to keep in mind for the future though...

File: Source/Core/VideoCommon/GraphicsModSystem/Runtime/CustomAssetLoader.cpp (hunk 1 above)
Comment (iwubcode) [reply Source/Core/VideoCommon/GraphicsModSystem/Runtime/CustomAssetLoader.cpp:21]: Yes it is copied.  The _hope_ is that the custom texture logic will end up using this functionality (both so that it can gain the asynchronous features this provides and so that we have one place for all texture operations).  If not, I'll probably have to expose some different way to set max memory size.

File: Source/Core/VideoCommon/GraphicsModSystem/Runtime/CustomAssetLibrary.cpp [hunk 2]
```diff
@@ -0,0 +1,154 @@
...
+
+namespace VideoCommon
+{
+bool CustomAssetLibrary::LoadGameTexture(const std::filesystem::path& asset_id,
+                                         CustomTextureData* data)
+{
+  if (!LoadTexture(asset_id, data))
+    return false;
+
+  const auto& first_mip = data->m_levels[0];
```
[the remark this answers is not in the document]
Comment (iwubcode) [reply Source/Core/VideoCommon/GraphicsModSystem/Runtime/CustomAssetLibrary.cpp:35]: It should be guaranteed.  I added some comments.  Apologies it wasn't there originally.

**Comment by iwubcode** [comment]
> I think you need to add a few comments here (in the code, please) to explain how you intend these classes to be used. I get the idea that you want to create the object at a defined place and then later asynchronously add data to it, but the details are not entirely clear from just these classes. In particular, the logic in the CustomAssetLoader is kinda hard to follow.

Thanks for the thorough review Admiral!

I've put comments all over, hopefully it is more obvious what things do.  Please let me know if there are any changes you'd like to see.

> What exactly is the asset_id, and why is it a std::filesystem::path? Does that make sense to be a path when we consider the possibility of synthesizing assets on the fly, or reading them from an archive without extracting/decompressing to disk?

AssetID is a unique identifier for an asset.  I was being dumb originally with it being a `std::filesystem::path`, it is now a `std::string`.

> I also feel like having a mutex per asset is probably not a great idea.

Would be curious to hear what you had in mind.  That was done intentionally so there's minimal contention between assets, as my understanding is locks are cheap when not being used.

---

AssetLoader now deals directly in terms of assets, it knows nothing of files.  I distinguished between a last loaded timestamp and a last written timestamp which removed the monitor struct that was there originally.  One point to mention about timestamps is I originally wanted to use `std::chrono::system_clock` but without `clock_cast` there's no way to convert from the filesystem times.  Given a filesystem is the only available library loader at this time, I just decided to leverage filesystem times with a `TODO` marking what we should do when `clock_cast` is available on all builders.

Otherwise just did some more general cleanup.  Please let me know if you see anything else.

**Comment by AdmiralCurtiss** [comment]
Can you (re)open the part you had here but not in #11871?
